src/shared/components/data-table/DataTable.test.tsx

```diff
@@ -0,0 +1,121 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import { render } from '@testing-library/react';
+import { times } from 'lodash';
+
+import DataTable from './DataTable';
+import { defaultDataTableState } from 'src/shared/components/data-table/dataTableReducer';
+
+import { type TableProps } from 'src/shared/components/data-table/DataTable';
+
+import { TableData, DataTableColumns } from './dataTableTypes';
+
+export const createDataTableSampleData = (
+  rows: number,
+  columns: number
+): { data: TableData; columns: DataTableColumns } => {
+  return {
+    data: times(rows, (row) =>
+      times(columns, (column) => `Cell ${row},${column}`)
+    ),
+    columns: times(columns, (column) => ({
+      columnId: `${column}`,
+      title: `Column ${column}`,
+      helpText: `Column ${column} help text`,
+      isSortable: true
+    }))
+  };
+};
+
+const { columns, data } = createDataTableSampleData(15, 10);
+
+const defaultProps = {
+  state: {
+    ...defaultDataTableState,
+    data
+  },
+  columns
+};
+
+describe('<DataTable />', () => {
+  const renderDataTable = (props: Partial<TableProps> = {}) =>
+    render(<DataTable {...defaultProps} {...props} />);
+
+  let container: HTMLElement;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('renders without error', () => {
+    container = renderDataTable().container;
+    expect(() => container).not.toThrow();
+  });
+
+  it('displays correct number of rows per page', () => {
+    container = renderDataTable({
+      state: {
+        ...defaultProps.state,
+        rowsPerPage: 10
+      }
+    }).container;
+
+    expect(container.querySelectorAll('tbody tr').length).toBe(10);
+
+    container = renderDataTable({
+      state: {
+        ...defaultProps.state,
+        rowsPerPage: 20
+      }
+    }).container;
+
+    expect(container.querySelectorAll('tbody tr').length).toBe(
+      defaultProps.state.data.length
+    );
+
+    container = renderDataTable({
+      state: {
+        ...defaultProps.state,
+        rowsPerPage: Infinity
+      }
+    }).container;
+
+    expect(container.querySelectorAll('tbody tr').length).toBe(
+      defaultProps.state.data.length
+    );
+
+    container = renderDataTable({
+      state: {
+        ...defaultProps.state,
+        rowsPerPage: 10,
+        currentPageNumber: 2
+      }
+    }).container;
+
+    expect(container.querySelectorAll('tbody tr').length).toBe(5);
+  });
+
+  it('respects onStateChange', async () => {
+    const onStateChange = jest.fn();
+    container = renderDataTable({
+      onStateChange
+    }).container;
+
+    expect(onStateChange).toBeCalledWith(defaultProps.state);
```

Comment:
I am not even seeing a state being changed here 🤔

Reply:
Here it just checks if `onStateChange` is called with the currentState when the table loads initially. 
I am not sure what is the best way to trigger a change and then test it.

Reply:
> Here it just checks if onStateChange is called with the currentState when the table loads initially.

That's my point: there was no change, but the onStateChange callback was triggered.

Reply:
I think so. Something native like an input element or a textarea element don't fire a change event when they are rendered with a provided value.

I wonder what should happen if the parent is in control of the state and decides to change it. The table probably shouldn't tell the parent back that the state has changed either; because the parent already knows it.

Reply:
I have pushed an update to call `onStateChange` only on state change